### ISSUE TITLE
build failure in macFileSystemWatcher.cpp with fmt

### ISSUE THREAD

**Issue by nieder** [MiKTeX/miktex#1335]
- [x] I have read and acknowledged this HOWTO article:  
  [Submitting a MiKTeX problem report](https://miktex.org/howto/bug-report)

Building miktex-23.5 on macOS, the build fails with this error:
> cd /sw/build.build/miktex-tools-23.5-1/miktex-23.5/finkbuild/Libraries/MiKTeX/Core/shared && /sw/var/lib/fink/path-prefix-clang/c++ -DFMT_SHARED -DHAVE_CONFIG_H -DHAVE_LIBZ=1 -DHAVE_ZLIB=1 -DHAVE_ZLIB_COMPRESS2=1 -DHAVE_ZLIB_H=1 -DMIKTEX -DMIKTEX_CORE_SHARED -DMIKTEX_MD5_SHARED -DMIKTEX_TRACE_SHARED -DMIKTEX_UNIX -DMIKTEX_UTIL_SHARED -Dmiktex_core_EXPORTS -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/finkbuild/Libraries/MiKTeX/Core/shared -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/MiKTeX/Core -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/finkbuild/Libraries/MiKTeX/Core -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/MiKTeX/UI/include -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/include -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/finkbuild/include -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/finkbuild/Libraries/MiKTeX/Core/include -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/MiKTeX/Core/include -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/3rd/fmt/source/include -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/finkbuild/Libraries/MiKTeX/Configuration/include -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/MiKTeX/Configuration/include -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/finkbuild/Libraries/MiKTeX/Util/include -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/MiKTeX/Util/include -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/finkbuild/Libraries/3rd/md5/include -I/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/MiKTeX/Trace/include -isystem /sw/include -MD -O3 -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.14 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -std=gnu++17 -MD -MT Libraries/MiKTeX/Core/shared/CMakeFiles/miktex-core.dir/__/FileSystemWatcher/mac/macFileSystemWatcher.cpp.o -MF CMakeFiles/miktex-core.dir/__/FileSystemWatcher/mac/macFileSystemWatcher.cpp.o.d -o CMakeFiles/miktex-core.dir/__/FileSystemWatcher/mac/macFileSystemWatcher.cpp.o -c /sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/MiKTeX/Core/FileSystemWatcher/mac/macFileSystemWatcher.cpp

```c++
In file included from /sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/MiKTeX/Core/FileSystemWatcher/mac/macFileSystemWatcher.cpp:27:
/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/MiKTeX/Core/internal.h:188:50: warning: 'RSA_free' is deprecated [-Wdeprecated-declarations]
using RSA_ptr = std::unique_ptr<RSA, decltype(&::RSA_free)>;
                                                 ^
/sw/include/openssl/rsa.h:293:1: note: 'RSA_free' has been explicitly marked deprecated here
OSSL_DEPRECATEDIN_3_0 void RSA_free(RSA *r);
^
/sw/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
#   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
                                                ^
/sw/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#     define OSSL_DEPRECATED(since) __attribute__((deprecated))
                                                   ^
In file included from /sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/MiKTeX/Core/FileSystemWatcher/mac/macFileSystemWatcher.cpp:24:
In file included from /sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/3rd/fmt/source/include/fmt/format.h:48:
/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/3rd/fmt/source/include/fmt/core.h:1735:47: error: cannot pass object of non-trivial type 'const MiKTeX::Util::PathName' through variadic
      method; call will abort at runtime [-Wnon-pod-varargs]
  const auto& arg = arg_mapper<Context>().map(FMT_FORWARD(val));
                                              ^
/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/3rd/fmt/source/include/fmt/core.h:204:26: note: expanded from macro 'FMT_FORWARD'
#define FMT_FORWARD(...) static_cast<decltype(__VA_ARGS__)&&>(__VA_ARGS__)
                         ^
/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/3rd/fmt/source/include/fmt/core.h:1777:10: note: in instantiation of function template specialization
      'fmt::v9::detail::make_value<fmt::v9::basic_format_context<fmt::v9::appender, char>, const MiKTeX::Util::PathName &>' requested here
  return make_value<Context>(val);
         ^
/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/3rd/fmt/source/include/fmt/core.h:1899:23: note: in instantiation of function template specialization 'fmt::v9::detail::make_arg<true,
      fmt::v9::basic_format_context<fmt::v9::appender, char>, fmt::v9::detail::type::custom_type, const MiKTeX::Util::PathName &, 0>' requested here
        data_{detail::make_arg<
                      ^
/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/3rd/fmt/source/include/fmt/core.h:1918:10: note: in instantiation of function template specialization
      'fmt::v9::format_arg_store<fmt::v9::basic_format_context<fmt::v9::appender, char>, MiKTeX::Util::PathName>::format_arg_store<const MiKTeX::Util::PathName &>' requested here
  return {FMT_FORWARD(args)...};
         ^
/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/3rd/fmt/source/include/fmt/core.h:3206:28: note: in instantiation of function template specialization
      'fmt::v9::make_format_args<fmt::v9::basic_format_context<fmt::v9::appender, char>, const MiKTeX::Util::PathName &>' requested here
  return vformat(fmt, fmt::make_format_args(args...));
                           ^
/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/MiKTeX/Core/FileSystemWatcher/mac/macFileSystemWatcher.cpp:58:45: note: in instantiation of function template specialization
      'fmt::v9::format<const MiKTeX::Util::PathName &>' requested here
        trace_files->WriteLine("core", fmt::format("adding directory to watch list: {0}", p));
                                            ^
In file included from /sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/MiKTeX/Core/FileSystemWatcher/mac/macFileSystemWatcher.cpp:24:
In file included from /sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/3rd/fmt/source/include/fmt/format.h:48:
/sw/build.build/miktex-tools-23.5-1/miktex-23.5/Libraries/3rd/fmt/source/include/fmt/core.h:1756:3: error: static_assert failed due to requirement 'formattable' "Cannot format an argument. To
      make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt"
  static_assert(
  ^
1 warning and 2 errors generated.
```
The above failure is with the included fmt. However, I get the same `static_assert failed due to requirement 'formattable'` error using a locally installed fmt-9.1.0 and adding the following flags to cmake:
```
		-DFMT_INCLUDE_DIR:PATH=/sw/include \
		-DFMT_LIBRARY=/sw/lib/libfmt.dylib \
		-DUSE_SYSTEM_FMT:BOOL=ON \
```

(OT: I had to add `${FMT_INCLUDE_DIR}` to several CMakeLists.txt to have the build find `<fmt/fmt.h>` when using system-fmt)

**Comment by briha** [comment]
Replacing `p` by `p.ToNativeString()` does the trick, at least as a temporary workaround.

**Comment by nieder** [comment]
`p` -> `p.ToNativeString()` worked for me too.

**Comment by edocevoli** [comment]
Thank you, I think this has been fixed with the 23.10 source tarball.
